library/x509_create.c

```diff
@@ -86,62 +92,161 @@ static const char *x509_at_oid_from_name( const char *name, size_t name_len )
     return( cur->oid );
 }
 
+static int x509_dn_check_special_chars( const char **cur, const char *end )
+{
+    const char *c = *cur;
+    int ret = 0;
+
+    /* Check for valid escaped characters */
+    if( c == end || *c != ',' )
+    {
+        ret = MBEDTLS_ERR_X509_INVALID_NAME;
+    }
+
+    return( ret );
+}
+
+static int dn_store_named_data( mbedtls_asn1_named_data **head, const char *oid,
+                                char data[MBEDTLS_X509_MAX_DN_NAME_SIZE], char *d,
+                                const char **s, const char **cur, const char *end )
+{
+    const char *c = *cur;
+
+    if( d - data == 0 )
+    {
+        return( MBEDTLS_ERR_X509_INVALID_NAME );
+    }
+
+    if( mbedtls_asn1_store_named_data( head, oid, strlen( oid ),
+                                       (unsigned char *) data,
+                                       d - data ) == NULL )
+    {
+        return( MBEDTLS_ERR_X509_ALLOC_FAILED );
+    }
+
+    /*
+     * Skip white spaces
+     */
+    while( c < end && *(c + 1) == ' ' )
+        c++;
+
+    *s = c + 1;
+
+    return ( 0 );
+}
+
 int mbedtls_x509_string_to_names( mbedtls_asn1_named_data **head, const char *name )
 {
     int ret = 0;
+    /*
+     * s points to the current string name.
+     * c points to the current character in the string being parsed.
+     * d points to the end of the current parsed named data.
+     */
     const char *s = name, *c = s;
     const char *end = s + strlen( s );
     const char *oid = NULL;
-    int in_tag = 1;
     char data[MBEDTLS_X509_MAX_DN_NAME_SIZE];
     char *d = data;
+    x509_dn_parse_state dn_cur_state = dn_in_tag;
 
     /* Clear existing chain if present */
     mbedtls_asn1_free_named_data_list( head );
 
     while( c <= end )
     {
-        if( in_tag && *c == '=' )
+        switch( dn_cur_state )
         {
-            if( ( oid = x509_at_oid_from_name( s, c - s ) ) == NULL )
+            /*
+             *Currently parsing a DN tag.
+             */
+            case dn_in_tag:
             {
-                ret = MBEDTLS_ERR_X509_UNKNOWN_OID;
-                goto exit;
+                if( *c == '=' )
+                {
+                    if( ( oid = x509_at_oid_from_name( s, c - s ) ) == NULL )
+                    {
+                        ret = MBEDTLS_ERR_X509_UNKNOWN_OID;
+                        goto exit;
+                    }
+                    s = c + 1;
+                    d = data;
+                    if( *s == '\"' )
```

Comment:
This doesn't handle white spaces correctly: We should skip whitespaces before and after the equality sign. With the present code, something like `A= "B"` would be treated as an unquoted value.

Reply:
I'll add a test case for this as well